p2p/peer.go

```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
+			// need to reserve the eviction for this specific peer such that
+			// others can't claim it.
+			if m.options.MaxConnected > 0 &&
+				len(m.connected) >= int(m.options.MaxConnected) &&
+				!m.peerIsUpgrade(peer, ranked) {
+				return "", PeerAddress{}, nil
+			}
+
 			m.dialing[peer.ID] = true
-			return peer.ID, peer.Addresses[0], nil
+			return peer.ID, addressInfo.Address, nil
 		}
 	}
 	return "", PeerAddress{}, nil
 }
 
+// retryDelay calculates a dial retry delay using exponential backoff, based on
+// retry settings in PeerManagerOptions. If MinRetryTime is 0, this returns
+// MaxInt64 (i.e. an infinite retry delay, effectively disabling retries).
+func (m *PeerManager) retryDelay(peer *peerInfo, failures uint32) time.Duration {
+	if failures == 0 {
+		return 0
+	}
+	if m.options.MinRetryTime == 0 {
+		return time.Duration(math.MaxInt64)
+	}
+	delay := m.options.MinRetryTime * time.Duration(math.Pow(2, float64(failures)))
+	if peer.Persistent && m.options.MaxRetryTimePersistent > 0 {
+		// We have to check this separately, since we don't want to fall back to
+		// MaxRetryTime in cases where it is smaller than MaxRetryTimePersistent.
+		if delay > m.options.MaxRetryTimePersistent {
+			delay = m.options.MaxRetryTimePersistent
+		}
+	} else if m.options.MaxRetryTime > 0 && delay > m.options.MaxRetryTime {
+		delay = m.options.MaxRetryTime
+	}
+	delay += time.Duration(rand.Int63n(int64(m.options.FuzzRetryTime))) // nolint:gosec
+	return delay
+}
+
 // DialFailed reports a failed dial attempt. This will make the peer available
 // for dialing again when appropriate.
-func (m *peerManager) DialFailed(peerID NodeID, address PeerAddress) error {
+//
+// FIXME: This should probably delete or mark bad addresses/peers after some time.
+func (m *PeerManager) DialFailed(peerID NodeID, address PeerAddress) error {
 	m.mtx.Lock()
 	defer m.mtx.Unlock()
+
 	delete(m.dialing, peerID)
-	// FIXME: We need to track address quality statistics and exponential backoff.
+
+	peer, err := m.store.Get(peerID)
+	if err != nil || peer == nil { // Peer may have been removed while dialing, ignore.
+		return err
+	}
+	if addressInfo := peer.LookupAddressInfo(address); addressInfo != nil {
+		addressInfo.LastDialFailure = time.Now().UTC()
+		addressInfo.DialFailures++
+		return m.store.Set(peer)
+	}
 	return nil
 }
 
 // Dialed marks a peer as successfully dialed. Any further incoming connections
 // will be rejected, and once disconnected the peer may be dialed again.
-func (m *peerManager) Dialed(peerID NodeID, address PeerAddress) error {
+func (m *PeerManager) Dialed(peerID NodeID, address PeerAddress) error {
 	m.mtx.Lock()
 	defer m.mtx.Unlock()
 
+	delete(m.dialing, peerID)
+
+	if m.connected[peerID] {
+		return fmt.Errorf("peer %v is already connected", peerID)
+	}
+	if m.options.MaxConnected > 0 &&
+		len(m.connected) >= int(m.options.MaxConnected)+int(m.options.MaxConnectedUpgrade) {
```

Comment:
We have a few places where we check this exact conditional. Perhaps move it to an auxiliary function for better readability?

Reply:
Yeah, maybe once we stabilize all the logic -- for now I'd like to keep it in each place, since the code is still in flux and we may need to vary it in each place.